### ISSUE TITLE
Built-in html formatter generates a blank html report if scenarios have not finished yet completely

### ISSUE THREAD

**Issue by seleniumghost** [cucumber/common#2004]
### 👓 What did you see?

When running tests using built-in html formatter, it generates a blank HTML file *until all scenarios finished completely*. Viewing the source of the html, I see a bunch of css and license stuff.

### ✅ What did you expect to see?

HTML report to be generated should not be blank *after a scenario has been finished*

### 📦 Which tool/library version are you using?

cucumber 8.0.0
cucumber-html-formatter 19.2.0
ruby 3.1.1
Windows 10

### 🔬 How could we reproduce it?

1. Have more than one scenario in a test.
2. Execute cucumber --format HTML --out report.html
3. Open the html report *when test is still running* and html file is blank.

### 📚 Any additional context?

I'm picking up the issue mentioned in comment https://github.com/cucumber/cucumber-ruby/issues/1524#issuecomment-824267664 from https://github.com/cucumber/cucumber-ruby/issues/1524 as it shows up as unacceptable drawback compared with older versions, especially when having a lot of scenarios or long scenarios or even both of this.

written by marnen:

> Unfortunately, the way that’s done makes the Cucumber HTML report much less useful than it used to be. Back in the day, Cucumber used to use a similar approach to RSpec: it wrote straight HTML as scenarios ran, so that the file would render something useful in the browser before it was completely written. That makes it ideal for monitoring a long run with lots of scenarios.
> 
> Current Cucumber HTML output, though, renders nothing until the run is complete (I think because `window.CUCUMBER_MESSAGES` isn’t well-formed JSON/JS till the final closing bracket is written), which means I need to wait for the entire run to finish before I see _any_ output. Is there an HTML formatter for current Cucumber that doesn’t have this drawback? I haven’t found one, and I really miss the old RSpec-like behavior. (And if this is off-topic, I’ll raise a new issue or ask on the list.)

answered by mattwynne:
 
> In fact, this kind of live monitoring of a test run from a browser should now be easier to do, given that the results are streamed out as little `ndjson` packets. The problem would be getting that stream from your Cucumber process to your browser. It's a really cool idea, but yes it is off-topic for this thread I think :) If you're interested in pursuing it, let's create a new issue (in https://github.com/cucumber/cucumber) and discuss it there.

----
 
*This text was originally generated from a [template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/about-issue-and-pull-request-templates), then edited by hand. [You can modify the template here.](https://github.com/cucumber/.github/edit/main/.github/ISSUE_TEMPLATE/bug_report.md)*

**Comment by seleniumghost** [comment]
Well, just to avoid misunderstanding and to be competely clear on that, I do not expect process monitoring, but partial output when a part (scenario) is finished. This means I expect behaviour which existed in former Cucumber versions.

**Comment by ciaranmcnulty** [comment]
Just to check, is this in a case where cucumber is writing to the filesystem and you're opening the file in your web browser?

**Comment by aurelien-reeves** [comment]
I don't think the actual html-formatter would be able to show any progress. It aims to generate an all-inclusive static report once the execution is done.

Yes, at some point, it kinda shows some progress, but that was only a side effect of how it was built before.

What you are requesting would require a brand new dedicated formatter which would actually be able to process messages stream as they arrive and thus monitor the execution

**Comment by aslakhellesoy** [comment]
I think maybe this can be solved without too much effort. Rather than writing a `window.CUCUMBER_MESSAGES` array at the end of the run, we could write messages as they come in - wrapped in a JavaScript function call:

```html
<script>processMessage({...})</script>
<script>processMessage({...})</script>
<script>processMessage({...})</script>
```

Every time another line is added, the UI would update. We'd have to make sure it doesn't impact performance when a finished page is loaded, but I don't think that's hard to do either.

**Comment by aslakhellesoy** [comment]
Thanks for raising this issue @seleniumghost - I think fixing this would be an improvement. I might take a crack at it one day, if I can find some time.
